### ISSUE TITLE
Forecast more timesteps ahead

### ISSUE THREAD

**Issue by EricavanderSar** [rte-france/Grid2Op#424]
## Is your feature request related to a problem? Please describe.
I am interested in including forecast when training a RLagent. I would like to know the effect of including a forecast of >1 ts ahead. This is currently not possible.

Ex. This is what i do:
```python
import grid2op
env = grid2op.make("rte_case5_example",reward_class=L2RPNSandBoxScore, test=True)
obs = env.reset()
forecast_prod_p = obs.get_forecasted_inj()[0]
...
```
By default we get the forecast of the power injections of 1 ts ahead.

## Describe the solution you'd like
I would like to be able to use 
```
obs.get_forecasted_inj(time_step=n)
# where n >1
```

I was able to retrieve the entire forecast of an episode via:
```
env.chronics_handler.real_data.data.load_p_forecast
```
But this is not something the agent can properly use.

Somewhere in the code I saw this comment
```
"""By default, forecasts are only made 1 step ahead. We could change that. Do not hesitate to make a feature request (https://github.com/rte-france/Grid2Op/issues/new?assignees=&labels=enhancement&template=feature_request.md&title=) if that is necessary for you."""
```
so here I am :)

Do you think this is something you can and will include? (sooner or later?) hope to hear from you. Thank you in advance!

Erica

**Comment by BDonnot** [comment]
Hello,

First answer : yes grid2op will fully support multi time step ahead (and already does so for the develpment version that you can install following instructions there: https://pypi.org/project/Grid2Op/1.8.2.dev3/

Then, you have multiple ways to interact with the "multiple steps ahead forecast", as you can see for example here : https://grid2op.readthedocs.io/en/dev_1.8.2/model_based.html

**BUT** the problem is the data generation process. It's far from easy to generate "good looking multiple steps ahead forecasts". We will generate some for some environments, but not for all. So you might not be able to read it.

**BUT BUT** I will add the possibility to "allow the agent to access multiple steps ahead *perfect* forecast" at creation time. I hope to have finished when I will do the grid2op new version release.

**Comment by EricavanderSar** [comment]
Thank you BDonnot for your quick response. I will have a look at the development version you shared!

**Comment by EricavanderSar** [comment]
Hello BDonnot,

Unfortunately, I am still not able to use the forecast as I would like. 
What I would like is simply the forecast of the production and consumption (forecast injections) in the environment. For this, I don't need a complete simulation.

So I would like to use

```
import grid2op
env = grid2op.make("rte_case5_example",reward_class=L2RPNSandBoxScore, test=True)
obs = env.reset()
forecast_prod_p = obs.get_forecasted_inj(time_step=n)[0]
...
```

with n >1

I see there is a new function to retrieve more forecasted information (get_forecast_arrays() also shows maintenance), which is interesting, but still only 1 ts ahead.
Furthermore, I saw there is the possibility to create a forecast_env, and try out actions here. Though I don't see yet how this is different from using simulation?
I think simulation would be much more computational expensive compared to what I would like. 
Therefore, I would like to know if extending the functionality of the forecast injections to > 1 ts ahead is still something you intend to include?

Looking forward to your reply,
Thank you!

**Comment by BDonnot** [comment]
Hello,

I get what you want. Unfortunately no forecast have been generated for more than a step ahead for already available environments.

One environment with this feature will come later on (hopefully by the end of the month). It will be based on the IEEE 118.

An alternative is to use grid2op to generate these forecasts "on the fly" (note however on this case you will not have a lot of "good properties" between your forecasts - eg you will not have spatial correlation for the error, nor temporal correlation between two horizon's [when forecast is made at the same time]) for that you need to install the grid2op 1.8.2.dev4 version and have a look at the documentation for example here (for the general concept, I highly recommend you read it to understand what you are doing) 
https://beta-grid2op.readthedocs.io/en/latest/timeserie_handlers.html

And especially here (for the actual class to use) :
https://beta-grid2op.readthedocs.io/en/latest/timeserie_handlers.html#grid2op.Chronics.handlers.NoisyForecastHandler

And then you can get the forecasted injections from multiple steps ahead. 

Also one other way would be to make yourself some forecast. Using any kind of methods you want (any model from scikit learn would to to start) 

`NB`: the problem you are asking a solution for his highly complicated. You ask, for every step t to provide forecasts for a given windows t+1, t+2,..., t+h (so if your episode counts 1k steps `t` you need to generate, if h is 12, 12k data) 
Implicitly this means ensuring generation and load balancing in each forecast grid, having some sort of spatial correlation, and temporal correlation with respect to both t and h, making sure a powerflow can run for each of the forecasted grid, being consistant in the sens that pmin, pmax and ramps are not violated, that you don't get solar at night, that wind generation looks like a wind time series, that some cheaper generators are preferred to more expensive one if possible and probably a lot more "things like that" 

In itself could be the topic of at least a PhD I think 😉

**Comment by BDonnot** [comment]
Glad to hear this is working for you.

If you want a "constant" forecast you can use the "PersistenceHandler" (or a name similar).

For the warning you are correct to ignore it. this is something I did not thought about but yes at the end of an episode you cannot forecast after its end. This does not make sense. I'll try to find a more elegant way to deal with this. Not sure there is much to do :-/ (if you have an idea let me know)

**Comment by EricavanderSar** [comment]
Thank you for pointing out the PersistenceForecastHandler. However, this one is simply copying the last data for the next t+1 to t+h time steps. Indeed, assuming a completely constant forecast, which does not really add any value for me.

The way I meant by saying "pretend the forecast stays the same" is as follows. Suppose at time step t we have predicted t+1 up to t+h. Now at time step t+1 we only need to add the prediction of timestep t+h+1 to get the forecast (t+1)+1 - (t+1)+h.
This (logically) works with PerfectForecast, but this is (as mentioned) usually not available. For now it will work to test the effect on the performance of the agent. 

Next, I can use the Noisy forecast, but here I will indeed need to replace the prediction t+1 up to t+h, since it becomes more accurate as we move forward in time. Or I might change the noise so that it is just as accurate for t+h, so:
```
...
"gen_p_for_handler": NoisyForecastHandler("prod_p_forecasted", sigma=lambda x: 0.01),
...
```
In which case I could do it as described and only add the new forecast (t+1)+h at t+1.

Thank you again. If I think of something I will let you know ;)

**Comment by BDonnot** [comment]
The code mentionned in this issue are now merged and available in latest grid2op version (1.9.0). I close this issue :-)
